### PR TITLE
Merge remote-tracking branch 'origin/master' into feature/refactoring

### PR DESCRIPTION
**Pull request by gumruyanzh**
Merge remote-tracking branch 'origin/master' into feature/refactoring

### CODE REVIEW

File: src/main/java/com/shop/data/entity/ProductDetailEntity.java [hunk 1]
```diff
@@ -28,7 +28,7 @@
 
     @OneToOne(fetch = FetchType.LAZY)
     @PrimaryKeyJoinColumn
-    private ProductEntity car;
+    private ProductEntity product;
 
     @Column(name = "mileage",nullable = false)
     private long mileage;
```
Comment (TorosyanV) [inline src/main/java/com/shop/data/entity/ProductDetailEntity.java]: Remove mileage

File: src/main/java/com/shop/service/dataservice/product/ProductPageViewService.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import java.util.List;
+
+/**
+ * Created by zhirayrg on 3/9/2017.
+ */
+public interface ProductPageViewService {
+
+    void addView(ProductViewDto viewDto);
+
+    List<PageViewItem> getViewHistory(long productId);
```
Comment (TorosyanV) [inline src/main/java/com/shop/service/dataservice/product/ProductPageViewService.java]: Use DayCountPair not PageViewItem

File: src/main/java/com/shop/service/dataservice/product/ProductServiceImpl.java [hunk 3]
```diff
@@ -172,59 +167,66 @@ public ProductWithDetailDto getCarWithDetail(long id) throws ProductNotFoundExce
...
     }
 
     @Override
     @Transactional(readOnly = true)
     public ProductEntity getByIdAndUserId(Long id, Long userId) throws InvalidProductOwnerException {
-        ProductEntity carEntity = carRepository.findByIdAndUserIdAndDeletedIsFalse(id, userId);
-        if (carEntity == null)
-            throw new InvalidProductOwnerException(String.format("User with id %s can't edit car with id %s", userId, id));
+        ProductEntity productEntity = productRepository.findByIdAndUserIdAndDeletedIsFalse(id, userId);
+        if (productEntity == null)
```
Comment (TorosyanV) [inline src/main/java/com/shop/service/dataservice/product/ProductServiceImpl.java]: not if (productEntity == null)
use findByIdAndDeletedIsFalse, after that check owner

need something like  productEntity.getOwner().getId().equals(userId)

File: src/main/java/com/shop/service/dataservice/security/SecurityServiceImpl.java [hunk 4]
```diff
@@ -41,6 +42,9 @@
     @Autowired
     RandomGenerator randomGenerator;
```
Comment (TorosyanV) [inline src/main/java/com/shop/service/dataservice/security/SecurityServiceImpl.java]: private

File: src/main/java/com/shop/service/dataservice/security/SecurityServiceImpl.java [hunk 5]
```diff
@@ -60,6 +64,25 @@ public boolean hasRole(String role) {
...
+    public UserEntity getUserByResetToken(String token) throws InvalidTokenException {
+        TokenEntity tokenEntity = tokenRepository.findByTokenTypeAndVal(TokenType.PASSWORD_RESET, token);
+        if (tokenEntity == null)
+            throw new InvalidTokenException("Reset Token not found");
+        UserEntity userEntity = tokenEntity.getUser();
+
+        return userEntity;
+    }
+
+    @Override
```
Comment (TorosyanV) [inline src/main/java/com/shop/service/dataservice/security/SecurityServiceImpl.java:78]: @Transactinal

File: src/main/java/com/shop/service/dataservice/security/SecurityServiceImpl.java [hunk 6]
```diff
@@ -60,6 +64,25 @@ public boolean hasRole(String role) {
     }
 
     @Override
+    public UserEntity getUserByResetToken(String token) throws InvalidTokenException {
```
Comment (TorosyanV) [inline src/main/java/com/shop/service/dataservice/security/SecurityServiceImpl.java]: Transactional readonly

File: src/main/java/com/shop/service/dto/product/ProductWithDetailDto.java [hunk 7]
```diff
@@ -92,5 +88,8 @@ public void setRegion(RegionSimpleDto region) {
     }
 
 
-
+    @Override
+    public String getDescription() {
+        return String.format("User Message about this product is %s", getDetail().getUserMessage());
```
Comment (TorosyanV) [inline src/main/java/com/shop/service/dto/product/ProductWithDetailDto.java]: return valid message or auto generated message

File: src/main/java/com/shop/web/Dashboard/DashboardDataController.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    @Autowired
+    private UserAnalyticService userAnalyticService;
+
+
+
+    @Autowired
+    private ProductAnalyticService productAnalyticService;
+
+
+    @RequestMapping(value = "/user/analytics/{joinCode}", method = RequestMethod.GET)
```
Comment (TorosyanV) [inline src/main/java/com/shop/web/Dashboard/DashboardDataController.java]: Use @GetMapping

File: src/main/java/com/shop/web/Dashboard/DashboardDataController.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    public List<DayCountPair> getUsersByJoinCode(@PathVariable("joinCode") String joinCode) {
+
+        return userAnalyticService.operatorsRegisteredUsersHistory(joinCode);
+
+    }
+
+    @RequestMapping(value = "/user/analytics/facebookusers", method = RequestMethod.GET)
```
Comment (TorosyanV) [inline src/main/java/com/shop/web/Dashboard/DashboardDataController.java]: @GetMapping

File: src/main/java/com/shop/web/Dashboard/DashboardDataController.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    public List<DayCountPair> getFacebookUsers() {
+
+        return userAnalyticService.facebookRegisteredUsersHistory();
+
+    }
+
+    @RequestMapping(value = "/user/analytics/allusers", method = RequestMethod.GET)
```
Comment (TorosyanV) [inline src/main/java/com/shop/web/Dashboard/DashboardDataController.java]: @GetMapping

File: src/main/java/com/shop/web/api/CurrencyController.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+ */
+@RestController
+@RequestMapping(value = "/api")
+public class CurrencyController {
+
+
+    @Autowired
+    private CurrencyService currencyService;
+
+    @RequestMapping(value = "/currencies", method = RequestMethod.GET)
```
Comment (TorosyanV) [inline src/main/java/com/shop/web/api/CurrencyController.java]: @GetMapping

File: src/main/java/com/shop/web/api/ProductDetailController.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+    @Autowired
+    private ProductPageViewService productPageViewService;
+
+    @Autowired
+    UserService userService;
+
+
+
+    @RequestMapping(value = "/product/{id}/json", method = RequestMethod.GET)
```
Comment (TorosyanV) [inline src/main/java/com/shop/web/api/ProductDetailController.java]: @GetMapping

File: src/main/java/com/shop/web/api/ProductDetailController.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+        try {
+            productWithDetailDto = productService.getProductWithDetail(productId);
+        } catch (ProductNotFoundException e) {
+            return null;
+        }
+        return productWithDetailDto;
+    }
+
+
+    @RequestMapping(value = "/product/analytics/{id}", method = RequestMethod.GET)
```
Comment (TorosyanV) [inline src/main/java/com/shop/web/api/ProductDetailController.java]: @GetMapping

File: src/main/java/com/shop/web/api/RegionController.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+@RestController
+@RequestMapping(value = "/api")
+public class RegionController {
+
+
+    @Autowired
+    private RegionService regionService;
+    @Autowired
+    private CityService cityService;
+    @RequestMapping( value = "/regions",method = RequestMethod.GET)
```
Comment (TorosyanV) [inline src/main/java/com/shop/web/api/RegionController.java]: @GetMapping

File: src/main/java/com/shop/web/api/RegionController.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    public List<RegionSimpleDto> getRegions() {
+
+        List<RegionSimpleDto> regionSimpleDtos = regionService.getAll();
+        return regionSimpleDtos;
+    }
+
+    @RequestMapping(value = "regions/{id}/cities", method = RequestMethod.GET)
```
Comment (TorosyanV) [inline src/main/java/com/shop/web/api/RegionController.java]: @GetMapping

File: src/main/java/com/shop/web/api/RegionController.java [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    public List<CitySimpleDto> getCities(@PathVariable("id") Long regionId) {
```
Comment (TorosyanV) [inline src/main/java/com/shop/web/api/RegionController.java]: not id, need to be regionId

File: src/main/java/com/shop/web/controller/DetailController.java [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+    @Autowired
+    private ProductService productService;
+
+    @Autowired
+    private ProductPageViewService productPageViewService;
+
+    @Autowired
+    private UserService userService;
+
+    @GetMapping("/detail/{id}")
```
Comment (TorosyanV) [inline src/main/java/com/shop/web/controller/DetailController.java]: not id, rproductId

File: src/main/java/com/shop/web/controller/HomeController.java [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+    private UserService userService;
+
+    @Autowired
+    private ProductService productService;
+
+    @Autowired
+    private MailService mailService;
+
+
+    @RequestMapping("/")
```
Comment (TorosyanV) [inline src/main/java/com/shop/web/controller/HomeController.java:33]: @GetMapping

File: src/main/java/com/shop/web/controller/ProductController.java [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+            logger.error(e.getMessage(), e.getCause());
+            return "ERROR";
+        } catch (InvalidProductOwnerException e) {
+            logger.info(e.getMessage(), e.getCause());
+
+            return "redirect:/";
+        }
+    }
+
+    @PostMapping("/delete/{id}")
```
Comment (TorosyanV) [inline src/main/java/com/shop/web/controller/ProductController.java:131]: not id, productId

File: src/main/java/com/shop/web/viewmodel/PageViewItem.java [hunk 20]
```diff
@@ -0,0 +1,31 @@
+package com.shop.web.viewmodel;
+
+/**
+ * Created by zhirayrg on 3/9/2017.
+ */
+public class PageViewItem {
```
Comment (TorosyanV) [inline src/main/java/com/shop/web/viewmodel/PageViewItem.java]: Not need, use DayViewPair....

**Review by TorosyanV — CHANGES_REQUESTED** [review]

File: src/main/java/com/shop/service/dataservice/product/ProductPageViewService.java (hunk 2 above)
Comment (gumruyanzh) [reply src/main/java/com/shop/service/dataservice/product/ProductPageViewService.java]: done

File: src/main/java/com/shop/service/dataservice/product/ProductServiceImpl.java (hunk 3 above)
Comment (gumruyanzh) [reply src/main/java/com/shop/service/dataservice/product/ProductServiceImpl.java]: done

File: src/main/java/com/shop/service/dataservice/security/SecurityServiceImpl.java (hunk 6 above)
Comment (gumruyanzh) [reply src/main/java/com/shop/service/dataservice/security/SecurityServiceImpl.java]: done

File: src/main/java/com/shop/service/dataservice/security/SecurityServiceImpl.java (hunk 4 above)
Comment (gumruyanzh) [reply src/main/java/com/shop/service/dataservice/security/SecurityServiceImpl.java]: +

File: src/main/java/com/shop/service/dataservice/security/SecurityServiceImpl.java (hunk 5 above)
Comment (gumruyanzh) [reply src/main/java/com/shop/service/dataservice/security/SecurityServiceImpl.java:78]: +

File: src/main/java/com/shop/service/dto/product/ProductWithDetailDto.java (hunk 7 above)
Comment (gumruyanzh) [reply src/main/java/com/shop/service/dto/product/ProductWithDetailDto.java]: +

File: src/main/java/com/shop/web/Dashboard/DashboardDataController.java (hunk 8 above)
Comment (gumruyanzh) [reply src/main/java/com/shop/web/Dashboard/DashboardDataController.java]: +

File: src/main/java/com/shop/web/Dashboard/DashboardDataController.java (hunk 9 above)
Comment (gumruyanzh) [reply src/main/java/com/shop/web/Dashboard/DashboardDataController.java]: +

File: src/main/java/com/shop/web/Dashboard/DashboardDataController.java (hunk 10 above)
Comment (gumruyanzh) [reply src/main/java/com/shop/web/Dashboard/DashboardDataController.java]: +

File: src/main/java/com/shop/web/api/CurrencyController.java (hunk 11 above)
Comment (gumruyanzh) [reply src/main/java/com/shop/web/api/CurrencyController.java]: +

File: src/main/java/com/shop/web/api/ProductDetailController.java (hunk 12 above)
Comment (gumruyanzh) [reply src/main/java/com/shop/web/api/ProductDetailController.java]: +

File: src/main/java/com/shop/web/api/RegionController.java (hunk 14 above)
Comment (gumruyanzh) [reply src/main/java/com/shop/web/api/RegionController.java]: +

File: src/main/java/com/shop/web/api/RegionController.java (hunk 15 above)
Comment (gumruyanzh) [reply src/main/java/com/shop/web/api/RegionController.java]: +

**Pull request merged**